subprojects/core/src/main/java/org/gradle/internal/operations/notify/BuildOperationNotificationBridge.java

```diff
@@ -35,16 +40,45 @@
 
     private Listener operationListener;
     private final BuildOperationListenerManager buildOperationListenerManager;
+    private BuildOperationRecorder buildOperationRecorder;
 
-    BuildOperationNotificationBridge(BuildOperationListenerManager buildOperationListenerManager) {
+    BuildOperationNotificationBridge(BuildOperationListenerManager buildOperationListenerManager, BuildOperationRecorder buildOperationRecorder, GradleInternal gradleInternal) {
         this.buildOperationListenerManager = buildOperationListenerManager;
+        this.buildOperationRecorder = buildOperationRecorder;
+        // ensure we only store events for configuration phase to keep overhead small
+        // when build scan plugin is not applied
+        gradleInternal.rootProject(new Action<Project>() {
+            @Override
+            public void execute(Project project) {
+                project.afterEvaluate(new Action<Project>() {
+                    @Override
+                    public void execute(Project project) {
+                        if (BuildOperationNotificationBridge.this.buildOperationRecorder != null) {
+                            BuildOperationNotificationBridge.this.buildOperationRecorder.stop();
+                        }
+                    }
+                });
+            }
+        });
     }
 
     @Override
     public void registerBuildScopeListener(BuildOperationNotificationListener notificationListener) {
         if (operationListener == null) {
             operationListener = new Listener(notificationListener);
             buildOperationListenerManager.addListener(operationListener);
+            List<BuildOperationRecorder.RecordedBuildOperation> recordedBuildOperations = buildOperationRecorder.getRecordedEvents();
+            for (BuildOperationRecorder.RecordedBuildOperation storedEvent : recordedBuildOperations) {
+                BuildOperationDescriptor buildOperationDescriptor = storedEvent.buildOperation;
+                Object event = storedEvent.event;
+                if (event instanceof OperationStartEvent) {
+                    operationListener.started(buildOperationDescriptor, (OperationStartEvent) event);
+                } else {
+                    operationListener.finished(buildOperationDescriptor, (OperationFinishEvent) event);
+                }
+            }
+            buildOperationRecorder.stop();
```

Comment:
maybe to make this a little clearer that `getRecordedEvents` is destructive, maybe make it a one shot method:

`buildOperationRecorder.retrieveEventsAndStop()` or

`buildOperationRecorder.stopAndDrainEventsTo(operationListener)`